python/pyspark/sql/tests/connect/client/test_reattach.py

```diff
@@ -0,0 +1,124 @@
+        assert cls.spark is not None
+
+    @classmethod
+    def tearDownClass(cls):
+        try:
+            # Stopping Spark Connect closes the session in JVM at the server.
+            cls.spark = cls.connect
+            del os.environ["PYSPARK_NO_NAMESPACE_SHARE"]
+        finally:
+            super(SparkConnectReattachTestCase, cls).tearDownClass()
+
+    def test_release_sessions(self):
+        big_enough_query = "select * from range(1000000)"
+        query1 = self.connect.sql(big_enough_query).toLocalIterator()
+        query2 = self.connect.sql(big_enough_query).toLocalIterator()
+        query3 = self.connect.sql("select 1").toLocalIterator()
+
+        next(query1)
+        next(query2)
+
+        jvm = PySparkSession._instantiatedSession._jvm  # type: ignore[union-attr]
+        service = getattr(
+            getattr(
+                jvm.org.apache.spark.sql.connect.service,  # type: ignore[union-attr]
+                "SparkConnectService$",
+            ),
+            "MODULE$",
+        )
+
+        @eventually(catch_assertions=True)
+        def wait_for_requests():
+            self.assertEqual(service.executionManager().listExecuteHolders().length(), 2)
+
+        wait_for_requests()
+
+        # Close session
+        self.connect.client.release_session()
+        # Calling release session again should be a no-op.
+        self.connect.client.release_session()
+
+        @eventually(catch_assertions=True)
+        def wait_for_responses():
+            self.assertEqual(service.executionManager().listExecuteHolders().length(), 0)
+
+        wait_for_responses()
+
+        # query1 and query2 could get either an:
+        # OPERATION_CANCELED if it happens fast - when closing the session interrupted the queries,
+        # and that error got pushed to the client buffers before the client got disconnected.
+        # OPERATION_ABANDONED if it happens slow - when closing the session interrupted the client
+        # RPCs before it pushed out the error above. The client would then get an
+        # INVALID_CURSOR.DISCONNECTED, which it will retry with a ReattachExecute, and then get an
+        # INVALID_HANDLE.OPERATION_ABANDONED.
+
+        def check_error(q):
+            try:
+                list(q)  # Iterate all.
+            except Exception as e:  # noqa: F841
+                return e
+
+        e = check_error(query1)
+        self.assertIsNotNone(e, "An exception has to be thrown")
+        self.assertTrue(
+            "OPERATION_CANCELED" in str(e) or "INVALID_HANDLE.OPERATION_ABANDONED" in str(e)
+        )
+        e = check_error(query2)
+        self.assertIsNotNone(e, "An exception has to be thrown")
+        self.assertTrue(
+            "OPERATION_CANCELED" in str(e) or "INVALID_HANDLE.OPERATION_ABANDONED" in str(e)
+        )
+
+        # query3 has not been submitted before, so it should now fail with SESSION_CLOSED
+        e = check_error(query3)
```

Comment:
We deleted `query3`, didn't we?

Reply:
We did at https://github.com/apache/spark/commit/9c291e1165c145104becf69ecafcdba2914c29f1. That was because of the behaviour difference between JVM and Python gRPC libraries:
- JVM sends an early request when opening a response stream from a request (https://github.com/grpc/grpc-java/issues/10697)
- Python does not send an request when opening a response stream (generator).

So I preserved the old test case here.